src/icrar/leap-accelerate/algorithm/cuda/PhaseRotate.cu

```diff
@@ -57,104 +58,76 @@ namespace icrar
+    void RotateVisibilities(
+        Integration& integration,
+        MetaDataCudaHost& metadata,
+        const MVDirection& direction)
     {
-        throw std::runtime_error("not implemented"); //TODO
-    }
+        //using namespace std::literals::complex_literals;
+        auto& data = integration.data;
+        auto& uvw = integration.uvw;
+        auto parameters = integration.parameters;
 
-    std::pair<Matrix<double>, Vector<std::int32_t>> PhaseMatrixFunction(
-        const Vector<std::int32_t>& a1,
-        const Vector<std::int32_t>& a2,
-        int refAnt, bool map)
-    {
-        auto unique = std::set<std::int32_t>(a1.cbegin(), a1.cend());
-        unique.insert(a2.cbegin(), a2.cend());
-        int nAnt = unique.size();
-        if(refAnt >= nAnt - 1)
+        if(metadata.init)
         {
-            throw std::invalid_argument("RefAnt out of bounds");
-        }
-
-        Matrix<double> A = Matrix<double>(a1.size() + 1, icrar::ArrayMax(a1));
-        for(auto v : A)
-        {
-            v = 0;
+            //metadata['nbaseline']=metadata['stations']*(metadata['stations']-1)/2
+            
+            metadata.SetDD(direction);
+            metadata.SetWv();
+            // Zero a vector for averaging in time and freq
+            metadata.avg_data = Eigen::MatrixXcd(integration.baselines, metadata.GetConstants().num_pols);
+            metadata.init = false;
         }
+        metadata.CalcUVW(uvw);
 
-        Matrix<int> I = Matrix<int>(a1.size() + 1, a1.size() + 1);
-        for(auto v : I)
+        // loop over baselines
+        for(int baseline = 0; baseline < integration.baselines; ++baseline)
         {
-            v = 1;
-        }
+            // For baseline
+            const double pi = boost::math::constants::pi<double>();
+            double shiftFactor = -2 * pi * uvw[baseline].get()[2] - metadata.oldUVW[baseline].get()[2]; // check these are correct
+            shiftFactor = shiftFactor + 2 * pi * (metadata.GetConstants().phase_centre_ra_rad * metadata.oldUVW[baseline].get()[0]);
+            shiftFactor = shiftFactor -2 * pi * (direction.get()[0] * uvw[baseline].get()[0] - direction.get()[1] * uvw[baseline].get()[1]);
 
-        int k = 0;
+            if(baseline % 1000 == 1)
+            {
+                std::cout << "ShiftFactor for baseline " << baseline << " is " << shiftFactor << std::endl;
+            }
 
-        for(int n = 0; n < a1.size(); n++)
-        {
-            if(a1(IPosition(n)) != a2(IPosition(n)))
+            // Loop over channels
+            for(int channel = 0; channel < metadata.GetConstants().channels; channel++)
             {
-                if((refAnt < 0) || ((refAnt >= 0) & ((a1(IPosition(n)) == refAnt) || (a2(IPosition(n)) == refAnt))))
+                double shiftRad = shiftFactor / metadata.GetConstants().channel_wavelength[channel];
+                double rs = sin(shiftRad);
+                double rc = cos(shiftRad);
```

Comment:
`std::sin` / `std::cos`. Won't make a difference now, but would if you decided to changed the type of `shiftRad`.

Reply:
rs and rc are unused variables from the python code. It looks like they were replaced with the complex number multiply so I'll remove them